enhancements/machine-api/control-plane-machineset-vertical-scaling.md

```diff
@@ -0,0 +1,365 @@
+
+## Summary
+
+To allow control plane nodes to remain at an adequate size for its cluster, this
+proposal introduces new configuration for the control plane machineset operator
+to allow it to make automated scaling decisions for control plane node sizing.
+
+## Motivation
+
+During normal operation of an OpenShift cluster, as worker node count and
+workloads increase, it becomes necessary to increase resources available to the
+control plane nodes.
+
+During SRE operations of the managed OpenShift product, there was already a
+trigger indentified, when these increases are required: right now these triggers
+are based on cluster load over the last 8 hours (details care available in [this
+PrometheusAlert](https://github.com/openshift/managed-cluster-config/blob/master/deploy/sre-prometheus/100-control-plane-resizing.PrometheusRule.yaml#L93C1-L93C1)
+that is used in OSD clusters). In addition to the average load our [recommended
+control plane
+practices](https://docs.openshift.com/container-platform/4.13/scalability_and_performance/recommended-performance-scale-practices/recommended-control-plane-practices.html)
+specify control plane sizes based on worker nodes.
+
+Instead of requiring the adjustments to be performed by hand, the operator
+should be able to automatically determine if a control plane node size change is
+required.
+
+On the other hand, it should also be possible to determine that control plane
+nodes have become too large for a cluster, because load has decreased. In this
+case the operator should also be able to automatically decide to scale the
+control plane down again.
+
+With the automatic increase of the control plane size, cluster performance can
+remain high, while automated decrease of the size will ensure that spend for
+users in a cloud environment is only has high as needed.
+
+### User Stories
+
+* As an OpenShift administrator, I want to know that the platform will always
+  have enough resources available to it's control plane nodes without manual
+  intervention.
+* As an OpenShift administrator, I want to know that the platform can reduce
+  spending on control plane nodes if they are oversized for the current cluster
+  load.
+* As an OpenShift administrator, I want to be able to control when sizing up and
+  down occurs, to adjust it to workloads on my cluster that might be burstable
+  to prevent the automatic scaling from occurring to often.
+
+### Goals
+
+* Allow automatic vertical scaling of control plane nodes.
+* Allow users to configure the sizes that can be chosen for automatic scaling.
+* Allow users to configure the thresholds when scaling will occur for CPU and
+  memory usage.
+* Allow users to configure how often checks for automatic scaling should be run.
+* Allow users to configure how long after a scale up or scale down, no more
+  scaling should be performed.
+* Allow users to completely disable scale up.
+* Allow users to completely disable scale down.
```

Comment:
In the interest of simplicity, these two seem like they might be covered by "Allow users to configure the sizes that can be chosen for automatic scaling.".  If your control plane is running type-b, you could drop the instance type list from [type-a, type-b, type-c] to just type-b to pin the cluster at the current type.  Or you could raise the "Allow users to configure how long after a scale up or scale down, no more scaling should be performed." threshold to infinity.  Perhaps there are additional constraints, and these are "...without having to check to see what instance the control plane is currently using"?  Or "...and disabling should pause any in-progress CPMS roll-outs"?  Or perhaps these entries can be dropped, and the other existing goals are sufficient?

Reply:
I was thinking of keeping these in, for the users that might never want to automatically scale down and only up for example.
I am not sure if that edge case is enough to have this out of the box, because generally I like the idea of having less configuration variables.

Reply:
I agree that there is separate value in disabling scale up/scale down. Having to remove the configuration of the different machine profiles to disable scale up/down feels like a bit of an awkward way to hold the tool, and it's not what we do in other autoscaling APIs to my knowledge

Reply:
I moved the whole decision making logic to scale up and down to what Custom Metrics Operator calls `triggers`.
with this approach, specifying no triggers would disable the respective option (scale up or scale down) completely.
I moved to the trigger approach, because CMO also allows the metrics API and prometheus as a source to trigger scaling, which is exactly what we want here as well. I tried to keep as many property as possible the same, so users of one of the operators might have an easy time to also setup the other.